emc/model.py

```diff
@@ -0,0 +1,216 @@
+"""A factory class that adapts DIPY's dMRI models."""
+import warnings
+import numpy as np
+from dipy.core.gradients import gradient_table
+
+
+class ModelFactory:
+    """A factory for instantiating diffusion models."""
+
+    @staticmethod
+    def init(gtab, model="TensorModel", **kwargs):
+        """
+        Instatiate a diffusion model.
+
+        Parameters
+        ----------
+        gtab : :obj:`numpy.ndarray`
+            An array representing the gradient table in RAS+B format.
+        model : :obj:`str`
+            Diffusion model.
+            Options: ``"3DShore"``, ``"SFM"``, ``"Tensor"``, ``"S0"``
+
+        Return
+        ------
+        model : :obj:`~dipy.reconst.ReconstModel`
+            An model object compliant with DIPY's interface.
+
+        """
+        if model.lower() in ("s0", "b0"):
+            return TrivialB0Model(gtab=gtab, S0=kwargs.pop("S0"))
+
+        # Generate a GradientTable object for DIPY
+        gtab = _rasb2dipy(gtab)
+        param = {}
+
+        if model.lower().startswith("3dshore"):
+            from dipy.reconst.shore import ShoreModel as Model
+
+            param = {
+                "radial_order": 6,
+                "zeta": 700,
+                "lambdaN": 1e-8,
+                "lambdaL": 1e-8,
+            }
+
+        elif model.lower().startswith("sfm"):
+            from emc.utils.model import (
+                SFM4HMC as Model,
+                ExponentialIsotropicModel,
+            )
+
+            param = {
+                "isotropic": ExponentialIsotropicModel,
+            }
+
+        elif model.lower().startswith("tensor"):
+            Model = TensorModel
+
+        elif model.lower().startswith("dki"):
+            Model = DKIModel
+
+        else:
+            raise NotImplementedError(f"Unsupported model <{model}>.")
+
+        param.update(kwargs)
+        return Model(gtab, **param)
+
+
+class TrivialB0Model:
+    """
+    A trivial model that returns a *b=0* map always.
+
+    Implements the interface of :obj:`dipy.reconst.base.ReconstModel`.
+    Instead of inheriting from the abstract base, this implementation
+    follows type adaptation principles, as it is easier to maintain
+    and to read (see https://www.youtube.com/watch?v=3MNVP9-hglc).
+
+    """
+
+    __slots__ = ("_S0",)
+
+    def __init__(self, gtab, S0=None, **kwargs):
+        """Implement object initialization."""
+        if S0 is None:
+            raise ValueError("S0 must be provided")
+
+        self._S0 = S0
+
+    def fit(self, *args, **kwargs):
+        """Do nothing."""
+
+    def predict(self, gradient, **kwargs):
+        """Return the *b=0* map."""
+        return self._S0
+
+
+class TensorModel:
+    """A wrapper of :obj:`dipy.reconst.dti.TensorModel."""
+
+    __slots__ = ("_model", "_S0", "_mask")
+
+    def __init__(self, gtab, S0=None, mask=None, **kwargs):
+        """Instantiate the wrapped tensor model."""
+        from dipy.reconst.dti import TensorModel
```

Comment:
TensorModel of Dipy and TensorModel of emc.model might be confusing. What about `import dipy.reconst.dti` and then we call ` dipy.reconst.dti.TensorModel`?